acme/order.go

```diff
@@ -253,18 +254,30 @@ func (o *order) finalize(db nosql.DB, csr *x509.CertificateRequest, auth SignAut
 		return nil, ServerInternalErr(errors.Errorf("unexpected status %s for order %s", o.Status, o.ID))
 	}
 
-	// Validate identifier names against CSR alternative names //
-	csrNames := make(map[string]int)
-	for _, n := range csr.DNSNames {
-		csrNames[n] = 1
+	// RFC8555: The CSR MUST indicate the exact same set of requested
+	// identifiers as the initial newOrder request. Identifiers of type "dns"
+	// MUST appear either in the commonName portion of the requested subject
+	// name or in an extensionRequest attribute [RFC2985] requesting a
+	// subjectAltName extension, or both.
+	if csr.Subject.CommonName != "" {
+		csr.DNSNames = append(csr.DNSNames, csr.Subject.CommonName)
 	}
-	orderNames := make(map[string]int)
-	for _, n := range o.Identifiers {
-		orderNames[n.Value] = 1
+	csr.DNSNames = uniqueLowerNames(csr.DNSNames)
+	orderNames := make([]string, len(o.Identifiers))
+	for i, n := range o.Identifiers {
+		orderNames[i] = n.Value
 	}
-	if !reflect.DeepEqual(csrNames, orderNames) {
+	orderNames = uniqueLowerNames(orderNames)
+
+	// Validate identifier names against CSR alternative names.
+	if len(csr.DNSNames) != len(orderNames) {
 		return nil, BadCSRErr(errors.Errorf("CSR names do not match identifiers exactly"))
 	}
+	for i := range csr.DNSNames {
+		if csr.DNSNames[i] != orderNames[i] {
+			return nil, BadCSRErr(errors.Errorf("CSR names do not match identifiers exactly"))
```

Comment:
Can we print the csr and order names here e.g.

```
BadCSRErr(errors.Errorf("CSR names do not match order identifiers exactly; CSR names = %s, Order names = %s", csr.DNSNames, orderNames))
```

Reply:
done.